src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -326,6 +287,59 @@ private void FreeNativeResources()
             }
         }
 
+        // The JIT is allowed to arbitrarily extend the lifetime of locals, which may retain SocketAsyncContext references,
+        // indirectly preventing Socket instances to be finalized, despite being no longer referenced by user code.
+        // To avoid this, the event handling logic is delegated to a non-inlined processing method.
+        // See discussion: https://github.com/dotnet/runtime/issues/37064
+        // SocketEventHandler holds an on-stack cache of SocketAsyncEngine members needed by the handler method.
+        private struct SocketEventHandler
+        {
+            public Interop.Sys.SocketEvent* Buffer { get; }
+
+            private readonly ConcurrentDictionary<IntPtr, SocketAsyncContextWrapper> _handleToContextMap;
+            private readonly ConcurrentQueue<SocketIOEvent> _eventQueue;
+
+            public SocketEventHandler(SocketAsyncEngine engine)
+            {
+                Buffer = engine._buffer;
+                _handleToContextMap = engine._handleToContextMap;
+                _eventQueue = engine._eventQueue;
+            }
+
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            public bool HandleSocketEvents(int numEvents)
+            {
+                bool enqueuedEvent = false;
+                foreach (var socketEvent in new ReadOnlySpan<Interop.Sys.SocketEvent>(Buffer, numEvents))
+                {
+                    IntPtr handle = socketEvent.Data;
+
+                    if (_handleToContextMap.TryGetValue(handle, out SocketAsyncContextWrapper contextWrapper))
+                    {
+                        SocketAsyncContext context = contextWrapper.Context;
+
+                        if (context.PreferInlineCompletions)
+                        {
+                            context.HandleEventsInline(socketEvent.Events);
+                        }
+                        else
+                        {
+                            Interop.Sys.SocketEvents events = context.HandleSyncEventsSpeculatively(socketEvent.Events);
+                            
+                            if (events != Interop.Sys.SocketEvents.None)
+                            {
+                                var ev = new SocketIOEvent(context, events);
+                                _eventQueue.Enqueue(ev);
```

Comment:
nit: we don't need to explicitly null the local variable anymore, so we can inline it

```suggestion
                                _eventQueue.Enqueue(new SocketIOEvent(context, events));
```